[IDE] Refactoring: Expand a ternary operator into an if statement and vice-versa

romanroibu opened the pull request:
When selecting a ternary expression
```
let x = expression ? a : b
```
Suggest expanding to an if-else:
```
let x: Int
if expression {
    x = a
} else {
    x = b
}
```
And vice-versa.

Feedback is very much appreciated.

Resolves [SR-5794](https://bugs.swift.org/browse/SR-5794).

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
   return false;
 }
 
+struct ExpandableTernaryExprInfo {
```
Can we make two sub-class of ExpandableTernaryExprInfo? Currently the logic for PatternBindingDecl and AssignExpr seems to be intertwined.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+  ExpandableTernaryExprInfo Target;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>())
+    Target.setBinding(dyn_cast<PatternBindingDecl>(D));
```
Once we separate the logic for PatternBindingDecl and AssignExpr, we can use constructors to pass the decl or the expr without using `setX()` functions.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+    return true; //abort
+
+  auto NameRange = Target.getNameRange();
+  auto NameCharRange = Lexer::getCharSourceRangeFromSourceRange(SM, NameRange);
```
can we simplify the logic by putting `Lexer::getCharSourceRangeFromSourceRange` into the `getNameRange` function.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,366 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+  llvm::raw_svector_ostream OS(DeclBuffer);
+
+  std::string space(" ");
+  std::string newline("\n");
```
We can use `StringRef Space = " "` here. Also, It's good practice to have consistent name conventions, i.e. to have "Space" and "NewLine" Here.

nkcsgexi reviewed (commented):
@romanroibu Overall looks good! 👍 I've reviewed the first half for now. I think it's a generally good idea to have one commit for each refactoring action though.

romanroibu commented:
Thanks for the review, @nkcsgexi ! I'll commit the fixes and get back to you ;)

romanroibu commented:
@nkcsgexi I've updated the PR with the changes you suggested. Let me know what else can be improved 👍

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+public:
+  virtual ~ExpandableTernaryExprInfo() {}
+
+  virtual swift::IfExpr *getIf() = 0;
```
We don't need `swift::` here. Could you clean them?

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+  }
+
+  swift::Type getType() {
+    return swift::Type();
```
`return nullptr;` should be sufficient here.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+    if (auto Pattern = getNamePattern())
+      return Pattern->getType();
+
+    return swift::Type();
```
`return nullptr;` here similarly.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>()) {
+    auto Binding = dyn_cast<PatternBindingDecl>(D);
+    auto BindingInfo = new ExpandableBindingTernaryExprInfo(Binding);
+    return std::unique_ptr<ExpandableTernaryExprInfo>(BindingInfo);
```
can we replace the above two statements with `return llvm::make_unique<ExpandableTernaryExprInfo>(Binding);`? In this way, we don't have a single point where a pointer is exposed.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+    return nullptr;
+
+  if (auto D = Info.ContainedNodes[0].dyn_cast<Decl*>()) {
+    auto Binding = dyn_cast<PatternBindingDecl>(D);
```
Check if `Binding` is nullptr here and only create `ExpandableBindingTernaryExprInfo` when it is not. Therefore, we don't need to check whether `Binding` is null inside `ExpandableBindingTernaryExprInfo`.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1667,6 +1667,391 @@ bool RefactoringActionConvertStringsConcatenationToInterpolation::performChange(
+
+  if (auto E = Info.ContainedNodes[0].dyn_cast<Expr*>()) {
+    auto Assign = dyn_cast<AssignExpr>(E);
+    auto AssignInfo = new ExpandableAssignTernaryExprInfo(Assign);
```
Similarly, check if `Assign` is nullptr before creating `ExpandableAssignTernaryExprInfo`.

nkcsgexi replied:
e.g.
```
if (auto Assign = dyn_cast<AssignExpr>(E)) {
  return llvm::make_unique<>(...)...;
}
```

nkcsgexi reviewed (commented):
Looks much better! Though, I've some new comments; could you take a look?

romanroibu commented:
I've fixed the issues you pointed out 👍

CodaFi commented:
Before this gets merged, could you squash this patch down into maybe two or three larger commits that encompass each feature that got added here?

romanroibu committed:
Refactoring: Add expand ternary expression action

romanroibu committed:
Refactoring: Add convert to ternary expression action

romanroibu commented:
Hey @nkcsgexi, I've addressed your latest comments, and also squashed the changes into 2 commits. Let me know if there's anything else that needs to be improved. Thanks!

harlanhaskins commented:
This patch looks good to me as-is! Thanks for doing this! I just wanna get one more approval from @nkcsgexi and get one last CI pass.

@swift-ci please smoke test

romanroibu committed:
Fix ExpandTernaryExpr expected test outputs

romanroibu commented:
@nkcsgexi I've fixed one issue with the tests, could you please trigger them again?

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

romanroibu committed:
Fix expand-ternary-expr and convert-to-ternary-expr tests

romanroibu commented:
I think all the issues should be solved, could you run the tests again, please?

CodaFi commented:
@swift-ci please smoke test

romanroibu committed:
Fix expand-ternary-expr and convert-to-ternary-expr tests

nkcsgexi commented:
@swift-ci please smoke test

romanroibu committed:
Fix expand-ternary-expr test

nkcsgexi commented:
@swift-ci please smoke test

romanroibu commented:
@nkcsgexi hmm, seems like it's failing because of the indentation. I thought the indentation is updated after the refactoring is applied, no? Should I just remove the indentation from the `.expected` files?

nkcsgexi commented:
@romanroibu we do indent them when users are invoking refactorings through IDE. However, indentation is another operation orchestrated by source editor. In here, the expected change will only include changes from the refactoring, which doesn't include indentation : )

romanroibu committed:
Fix expand-ternary-expr test (remove indentation)

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@romanroibu Merged! Thank you for the awesome addition to Swift refactoring!

romanroibu commented:
Thank you for all the help! :)
